pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -0,0 +1,257 @@
+		// if len(split.AppendHeaders) > 0 {
+		// 	h = &v1alpha3.Headers{
+		// 		Request: &v1alpha3.HeaderOperations{
+		// 			Add: split.AppendHeaders,
+		// 		},
+		// 	}
+		// }
+
+		weights = append(weights, v1alpha3.HTTPRouteDestination{
+			Destination: v1alpha3.Destination{
+				Host: network.GetServiceHostname(
+					split.ServiceName, split.ServiceNamespace),
+				Port: makePortSelector(split.ServicePort),
+			},
+			Weight:  split.Percent,
+			Headers: h,
+		})
+	}
+
+	var h *v1alpha3.Headers
+	// TODO(mattmoor): Switch to Headers when we can have a hard
+	// dependency on 1.1, but 1.0.x rejects the unknown fields.
+	// if len(http.AppendHeaders) > 0 {
+	// 	h = &v1alpha3.Headers{
+	// 		Request: &v1alpha3.HeaderOperations{
+	// 			Add: http.AppendHeaders,
+	// 		},
+	// 	}
+	// }
+
+	return &v1alpha3.HTTPRoute{
+		Match:   matches,
+		Route:   weights,
+		Timeout: http.Timeout.Duration.String(),
+		Retries: &v1alpha3.HTTPRetry{
+			Attempts:      http.Retries.Attempts,
+			PerTryTimeout: http.Retries.PerTryTimeout.Duration.String(),
+		},
+		// TODO(mattmoor): Remove AppendHeaders when 1.1 is a hard dependency.
+		// AppendHeaders is deprecated in Istio 1.1 in favor of Headers,
+		// however, 1.0.x doesn't support Headers.
+		DeprecatedAppendHeaders: http.AppendHeaders,
+		Headers:                 h,
+		WebsocketUpgrade:        true,
+	}
+}
+
+func dedup(hosts []string) []string {
+	return sets.NewString(hosts...).List()
+}
+
+func retainLocals(hosts []string) []string {
+	localSvcSuffix := ".svc." + network.GetClusterDomainName()
+	retained := []string{}
+	for _, h := range hosts {
+		if strings.HasSuffix(h, localSvcSuffix) {
+			retained = append(retained, h)
+		}
+	}
+	return retained
+}
+
+func expandedHosts(hosts []string) []string {
+	expanded := []string{}
+	allowedSuffixes := []string{
+		"",
+		"." + network.GetClusterDomainName(),
+		".svc." + network.GetClusterDomainName(),
+	}
+	for _, h := range hosts {
+		for _, suffix := range allowedSuffixes {
+			if strings.HasSuffix(h, suffix) {
+				expanded = append(expanded, strings.TrimSuffix(h, suffix))
+			}
+		}
+	}
+	return dedup(expanded)
+}
+
+func makeMatch(host string, pathRegExp string) v1alpha3.HTTPMatchRequest {
+	match := v1alpha3.HTTPMatchRequest{
+		Authority: &istiov1alpha1.StringMatch{
+			Regex: hostRegExp(host),
+		},
+	}
+	// Empty pathRegExp is considered match all path. We only need to
+	// consider pathRegExp when it's non-empty.
+	if pathRegExp != "" {
+		match.URI = &istiov1alpha1.StringMatch{
+			Regex: pathRegExp,
+		}
+	}
+	return match
+}
+
+// Should only match 1..65535, but for simplicity it matches 0-99999.
+const portMatch = `(?::\d{1,5})?`
+
+// hostRegExp returns an ECMAScript regular expression to match either host or host:<any port>
+func hostRegExp(host string) string {
+	return fmt.Sprintf("^%s%s$", regexp.QuoteMeta(host), portMatch)
+}
+
+func getHosts(ci *v1alpha1.Ingress) []string {
```

Comment:
This only depends on `ci.Spec`, so may be we can share with ClusterIngress controller.

Reply:
Done.